src/libexpr/primops.cc

```diff
@@ -705,38 +705,72 @@ static RegisterPrimOp primop_genericClosure(PrimOp {
-      named `key` which is comparable. The result is produced by calling `operator`
-      for each `item` with a value for `key` that has not been called yet including
-      newly produced `item`s. The function terminates when no new `item`s are
-      produced. The resulting *list of attrsets* contains only *attrsets* with a
-      unique key. For example,
+      `builtins.genericClosure` computes the transitive closure over an arbitrary relation defined by an `operator` function.
+      It is designed to handle potentially cyclic data structures or any complex language constructs that require iterative processing.
 
-      ```
-      builtins.genericClosure {
-        startSet = [ {key = 5;} ];
-        operator = item: [{
-          key = if (item.key / 2 ) * 2 == item.key
-               then item.key / 2
-               else 3 * item.key + 1;
-        }];
-      }
-      ```
-      evaluates to
-      ```
-      [ { key = 5; } { key = 16; } { key = 8; } { key = 4; } { key = 2; } { key = 1; } ]
-      ```
+      Common usages are:
+
+      - Generating dependency trees or unique collections of items.
+      - Traversing through structures that may contain cycles or loops.
+      - Processing data structures with complex relationships.
+
+      ## Concept
+
+      - **startSet**: The initial collection of attribute sets from which the closure begins its computation.
+      - **operator**: A function that takes an attrset and returns a list of attribute sets. It defines how each item in the current set is processed and expanded into more items.
+
+      The result is produced by calling the `operator` recursively over again, until no new items (compared by key) are returned.
+
+      The `genericClosure` function terminates when no new `items` are produced.
+
+      > **Example**
+      >
+      > ```nix
+      > builtins.genericClosure {
+      >   startSet = [ {key = 5;} ];
+      >   operator = item: [{
+      >     key = if (item.key / 2 ) * 2 == item.key
+      >          then item.key / 2
+      >          else 3 * item.key + 1;
+      >   }];
+      > }
+      > ```
+      >
+      > evaluates to
+      >
+      > ```nix
+      > [ { key = 5; } { key = 16; } { key = 8; } { key = 4; } { key = 2; } { key = 1; } ]
+      > ```
+
+      > **Prerequisites**
+      >
+      > - Ensure key in startSet and operator's output is of the type listed below.
+      > - Repeated application of the operator function to the working set must eventually stop producing new keys.
+
+      ## Reference
 
       `key` can be one of the following types:
+
       - [Number](@docroot@/language/values.md#type-number)
       - [Boolean](@docroot@/language/values.md#type-boolean)
       - [String](@docroot@/language/values.md#type-string)
       - [Path](@docroot@/language/values.md#type-path)
       - [List](@docroot@/language/values.md#list)
 
+      > **Type** (pseudocode)
+      >
+      > ```nix
+      > genericClosure :: {
```

Comment:
We don't use this pseudo-Haskell type notation anywhere else in the manual, so this may raise more questions than it answers for people not familiar with Haskell.

Reply:
They're very common elsewhere in the Nix ecosystem though, including Nixpkgs. Providing a link or short explainer on how to read that notation somewhere would be a better option than the current inconsistency.

Reply:
The Nix manual has too much authority for using something as ill-defined and flawed. We'd accidentally make the syntax official. This is not something we can do without making sure that the type language is good:
- unambiguous syntax
- unambiguous semantics, ie actually checkable
- not overly complicated (Haskell `[t]` was a mistake)

Standardized formatting would also be helpful. The indentation in this signature does not look optimized for understanding.

Reply:
Ack. Had the same thoughts when writing this. Removed the 'type' section.